### PR TITLE
8333925: Shenandoah: Heuristics should have an option to ignore abbreviated cycles

### PR DESCRIPTION
**Pull request by earthling-amzn**
After concurrent marking is complete, Shenandoah will skip the evacuation phase if a sufficient amount of garbage is found in regions that contain _no live objects_. These abbreviated cycles are much shorter than a cycle that performs evacuation and update references and tend to lower the average cycle time used by the heuristic to predict cycle times. This may cause the heuristic to wait too long to initiate a cycle and may lead to degenerated cycles. This change has the heuristic ignore abbreviated cycle times by default, with an option to have the heuristic count them as it does now.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333925](https://bugs.openjdk.org/browse/JDK-8333925): Shenandoah: Heuristics should have an option to ignore abbreviated cycles (**Task** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Author) ⚠️ Review applies to [d8036f31](https://git.openjdk.org/jdk/pull/19640/files/d8036f316a69bad162923707552527b64fa27c79)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - **Reviewer**) ⚠️ Review applies to [d8036f31](https://git.openjdk.org/jdk/pull/19640/files/d8036f316a69bad162923707552527b64fa27c79)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19640/head:pull/19640` \
`$ git checkout pull/19640`

Update a local copy of the PR: \
`$ git checkout pull/19640` \
`$ git pull https://git.openjdk.org/jdk.git pull/19640/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19640`

View PR using the GUI difftool: \
`$ git pr show -t 19640`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19640.diff">https://git.openjdk.org/jdk/pull/19640.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19640#issuecomment-2159427095)

### CODE REVIEW

**Review by kdnilsen — APPROVED** [review]

**Review by ysramakrishna — APPROVED** [review]
LGTM.

Any performance numbers or improvements to share?

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp [hunk 1]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2018, 2020, Red Hat, Inc. All rights reserved.
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp]: Okay.

File: src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp [hunk 2]
```diff
@@ -149,6 +149,16 @@
           "Larger values give more weight to recent values.")               \
           range(0,1.0)                                                      \
                                                                             \
+  product(bool, ShenandoahAdaptiveIgnoreShortCycles, true, EXPERIMENTAL,    \
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp]: Good idea.

**Comment by earthling-amzn** [comment]
Withdrawing this PR because I cannot reproduce the results that compelled us to make this change in the `shenandoah` repo.

**Pull request closed without merging**